### PR TITLE
[JENKINS-28289] Ensure proxy exclusion list is used while getting proxy for JNLP clients

### PR DESCRIPTION
**Pull request by evanx9**
Pull request as planned : https://issues.jenkins-ci.org/browse/JENKINS-28289

Ensure proxy exclusion list is used while getting proxy for JNLP
clients. Updated Jacoco version. Reworked version from
https://github.com/jenkinsci/remoting/pull/55

### CODE REVIEW

**Comment by evanx9** [comment]
Looks like something is wrong with the artifact repo? The build portion looks good. @daniel-beck

**Comment by evanx9** [comment]
@oleg-nenashev Noticed you are investigating the errors during deployment stage on another pull. Do let me know if the job can be fixed.

**Comment by daniel-beck** [comment]
@evanx9 Don't worry, I think we can safely ignore brokenness like this that's unrelated to your changes.

**Comment by oleg-nenashev** [comment]
retriggering the build

File: src/main/java/hudson/remoting/Engine.java [hunk 1]
```diff
@@ -351,8 +342,15 @@ private Socket connect(String host, String port) throws IOException, Interrupted
...
-                if(retry++>10)
-                    throw (IOException)new IOException("Failed to connect to "+host+':'+port).initCause(e);
+            	if (retry++>10) {
+            		String suffix = "";
+            		if (isHttpProxy) {
+            			suffix = " through proxy " + targetAddress.toString();
+            		}
+            		// Cast -> InitCause method is returning a type Throwable.
+            		// Therefore we need to cast it to IOException.
+            		throw (IOException) new IOException("Failed to connect to " + host + ':' + port + suffix).initCause(e);
```
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Engine.java]: NIT: I would decouple it to 3 lines and avoid any casts. BTW this approach should also work

File: src/main/java/hudson/remoting/Util.java [hunk 2]
```diff
@@ -131,4 +136,37 @@ static URLConnection openURLConnection(URL url) throws IOException {
...
+    
+    static InetSocketAddress getResolvedHttpProxyAddress(String host, int port) throws IOException {
+    	InetSocketAddress targetAddress = null;
+    	Iterator<Proxy> proxies = ProxySelector.getDefault().select(URI.create(String.format("http://%s:%d", host, port))).iterator();
+    	// No break -> The while condition already checks if targetAddress is null.
+    	while (targetAddress == null && proxies.hasNext()) {
+    		Proxy proxy = proxies.next();
+    		if (proxy.type() == Proxy.Type.HTTP) {
+    			if (!(proxy.address() instanceof InetSocketAddress))
+    				throw new IOException("Unsupported proxy address type " + proxy.address().getClass());
```
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Util.java]: <code>proxy.address</code> may return null if it is a direct connection. In such case there will be an NPE in the routine. 
If I understand the code correctly, <code> getResolvedHttpProxyAddress</code> should return null in such case to invoke the no-proxy logic in the calling method

File: src/main/java/hudson/remoting/Util.java [hunk 3]
```diff
@@ -131,4 +136,37 @@ static URLConnection openURLConnection(URL url) throws IOException {
...
+    		}
+    	}
+    	if (targetAddress == null) {
+    		String httpProxy = System.getenv("http_proxy");
+    		if (httpProxy != null) {
+    			try {
+    				URL url = new URL(httpProxy);
+    				targetAddress = new InetSocketAddress(url.getHost(), url.getPort());
+    			} catch (MalformedURLException e) {
+    				LOGGER.log(Level.WARNING, INVALID_HTTP_PROXY_MSG + " HTTP Proxy: " + httpProxy);
```
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Util.java]: Other calls from remoting or Jenkins slave may try to use this corrupted proxy setting. Probably it's safer to just throw <code>IOException</code> and fail here.

**Comment by oleg-nenashev** [comment]
@evanx9 , @hypnoce 
See my comments above.

I suppose jacoco update is unrelated. For the future, it's better to use separate pull-requests for such updates. If you also resubmit a commit from other developer, please also make sure that you retain the authorship. It may be sensitive to them. If you make a patch on the top of the commit, just keep it as a separate commit.

I also have minor comments regarding the formatting. The project uses a subset of [Oracle code style](http://www.oracle.com/technetwork/java/codeconvtoc-136057.html) 
* The newly added code has tab indenting instead of the space one
* There are pretty much single-line <code>if</code> bodies without brackets. Since TAB indenting differs, it may confuse other developers if the code gets aligned improperly
* Sometimes annotations are helpful for static analysis. If your method may return a null, it's a good practice to annotate it via <code>@CheckForNull</code>

**Comment by evanx9** [comment]
@oleg-nenashev Updated with your suggestions. Code is using spaces now. Noticed a problem with the testExportSecondary Test with Cloudbee's Jenkins. This test passes on my local system. I don't think my changes should have impacted this. Could you re-run the job to confirm?

**Comment by hypnoce** [comment]
Is it possible to keep my commits in the git log ? It does not appear anymore. Thanks.

**Comment by daniel-beck** [comment]
@evanx9 Please create commits that have @hypnoce as author if you're reusing his work.

@oleg-nenashev Time for another review?

**Comment by daniel-beck** [comment]
Retriggering PR build.

**Comment by evanx9** [comment]
@daniel-beck My commit messages clearly site both hypnoce and his original pull request, and is a modification on top of his changes. would that be sufficient?

The new build looks clean. Looking forward to merges with master and potentially the LTS, as we are having issues with the proxy and exclusion list right now with new nodes.

**Comment by oleg-nenashev** [comment]
I'll try to review it tomorrow

**Comment by hypnoce** [comment]
PR #55  has been updated.

**Comment by oleg-nenashev** [comment]
#55 has been merged. I think this PR can be closed

**Pull request closed without merging**